### ISSUE TITLE
[Master] GDScript raises an error when a node name is an integer and you try to fetch it via a script

### ISSUE THREAD

**Issue by xsellier** [godotengine/godot#55088]
### Godot version

4.0.dev 15062513c01a835f59d9c905e5e65a7002d6c370

### System information

Linux binogure 5.14.0-2-amd64 #1 SMP Debian 5.14.9-2 (2021-10-03) x86_64 GNU/Linux

### Issue description

GDScript raises an error when a node name is an integer and you try to fetch it via a script
![image](https://user-images.githubusercontent.com/6616005/142424477-bb427210-2cf8-4c27-9bd7-d93dd35ae7a5.png)

### Steps to reproduce

Create a scene, and add any type of node as a child, then name that node `1`, attach a script to the main node, and try the following code:
```gdscript
func _ready:
  $1.get_name()
```

**Comment by ator-dev** [comment]
This seems intentional to me. There's a specific case in the parser to exclude literals (which I think just means 'identifiers' interpreted as ints?), and literals aren't in the list of types that work in a NodePath. Should this just be adapted to not exclude literals (and add them to the accepted list), or is there an important reason for this behaviour that means a special case should be added for some types of literals?

![image](https://user-images.githubusercontent.com/66553618/142759853-506340b3-8b3c-4d57-b54c-085b24b82e52.png)

**Comment by Ev1lbl0w** [comment]
It works if you use a string or `get_node` instead:
```gdscript
$"1".get_name()           # OK
get_node("1").get_name()  # OK
$1.get_name()             # Error
```
Since `$1` is supposed to be a shorthand for `get_node`, I'd say this is a bug.

**Comment by ator-dev** [comment]
Good point, I agree there isn't an obvious reason this shouldn't be allowed. There's no danger of the user's intention being something else or the literal getting interpreted wrongly, since the number would always be placed as part of a `/` separated node path after `$`. I'll make a pull request on changing this, and see if it works out or there are any objections.

**Comment by ator-dev** [comment]
Turns out it's (at least partly) because a `.` placed after an integer is interpreted as a decimal point, which obviously causes problems in parsing the code. Since this is a more fundamental limitation than I realised, I will have a look at it again later since there's quite a lot of other work to get done right now (;

**Comment by akien-mga** [comment]
This isn't really a bug IMO, and indeed it works with `$"1".name` so it's not a big deal.

But maybe support could be added for it if anyone wants to have a look.

**Comment by xsellier** [comment]
I aknowledge this might be a _simple_ corner case (there are more trickier corner cases for sure, #42410 #72973 #73273 ). With more GDScript user you will encounter more corner cases such this one.
Fixing all the corner cases will take a lot of time, but at the end it will make GDScript rock solid.
